heartbeat/heartbeat.py

```diff
@@ -23,5 +24,15 @@
         print("Variable TEMPLATE_ID is missing")
     for base_url in BASE_URL:
         notifications_client = NotificationsAPIClient(API_KEY, base_url=base_url)
-        response = notifications_client.send_email_notification(email_address=EMAIL_ADDRESS, template_id=TEMPLATE_ID)
-        print("Email has been sent!")
+        try:
+            response = notifications_client.send_email_notification(email_address=EMAIL_ADDRESS, template_id=TEMPLATE_ID)
+            print("Email has been sent by {}!".format(base_url))
+        except HTTPError as e:
+            print("There was an error")
+            print(e.status_code)
+            print(e._message)
+            raise
+        except Exception as e:
+            raise
```

Comment:
No need to except and raise: there is no error handling. Omitting this will have the same effect.

Reply:
I guess thats fair, it will just bubble up the stack. I just think this is cleaner and would like to leave it as such.

Reply:
Hmm no.. I will challenge you on that because it is not cleaner. But I would like to know why you think it is? Also, can you refer to me documented best practices, or usage of this style in the Python code base (or Notify)? Stackoverflow articles maybe? 

This is to me against many best practices I know, i.e. 1. don't write error handling code if we don't handle error and 2. write less code.

Reply:
That's fair I shouldn't have resolved, my bad.

This is my logic - when looking at a try, except -> you think where is this exception being raised? Logically you look at the `except` in the code but it won't be line 30 if it isn't HTTPError.  So then you go back to line 28 which is raising the exception, though by convention you don't want to think of the `try` raising the exception.

that being said, best practice is to never `except Exception` but catch specific `exceptions`.
Is this redundant code? Yes
Does it make it cleaner and more obvious though? Imo Yes

Regardless, I am not tied to one way to another and am open to changing it even if I prefer the other way.

Reply:
I think I get what you are saying (I think): that logic around raising exception is implicit and here you make it explicit. Hence it is more clear. Is that how you see it?

Reply:
yup that was a pretty concise summary of what I was trying to say.

Reply:
**TL;DR**: I have the impression here that what we might desire is stronger type safety for more explicit control flow.

Okay I see. In a way, I agree and I am in the camp that these things should be explicit (and hence I am a fan of languages with stronger type safety where types and control flow is explicit), but alas, Python is not such a programming language.

I think it's **not** a common practice in the Python community because the non explicit control flow is accepted (or even static typing until recent effort). In other communities like Haskell though, maximum expression of the control flow is desired, but that comes at a price of hitting walls against the compiler and extra types. Python takes a more loose approach and focus on getting things done (sorry Haskell, I still love you).

Consider runtime exceptions: these are dynamic. When a runtime exception occurs, they bubble up the stack and are not necessarily all advertised through the functions/methods signatures or even through all the encountered `try...except` clauses. This is a common criticism of using runtime/dynamic exceptions in these programming languages: they can occur everywhere and control flow is quite dynamic.

An empty error management block is not a common practice in such languages though (or any that I know of): imagine that every time we write a `try`, we will now add an `except Exception... raise` block with no error handling code, on the purpose of _clarity_ (or explicit declaration that something might happen but... 🤞 ). It's a losing game because these are all dynamic runtime exceptions and they can spawn easily everywhere.

Keeping an open mind: let's say we adopt this practice, we decide to go that way of defining explicitly `try... raise Exception`, should we add an empty error management block every time we do a `try...` block in Notify? Should we add this to existing blocks in the code base? all of our repos? Would it be really clearer and more explicit? I'd say it's a known behavior and normal for Python. We do a `try` if we think there is a code point that can handle an error usually, for fine (i.e. `HTTPError`) or large grained errors (i.e. `Exception` and its various problematic children classes). We don't do it **solely** to declare that something might happen.

Being more explicit in Python with dynamic runtime exceptions would mean more code cluttering of empty `except...raise` blocks. I can change my mind on that, but I'd like to see real world examples of this practice and documentation/discussion about it.

Sorry for the long message, this was not a concise summary that time.